src/main/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecution.java

```diff
@@ -559,34 +594,93 @@ int approximateNodeCount() {
             // heads could not be restored in unmarshal, so doing that now:
             heads = new TreeMap<Integer,FlowHead>();
             for (Map.Entry<Integer,String> entry : headsSerial.entrySet()) {
                 FlowHead h = new FlowHead(this, entry.getKey());
-                h.setForDeserialize(storage.getNode(entry.getValue()));
-                heads.put(h.getId(), h);
+
+                FlowNode n = storage.getNode(entry.getValue());
+                if (n != null) {
+                    h.setForDeserialize(storage.getNode(entry.getValue()));
+                    heads.put(h.getId(), h);
+                } else {
+                    // FlowNodeStorage not in synch with the FlowExecution -- hard failure of master with storage not persisted cleanly
+                    // Or files deleted (maybe build rotation).
+                    // Need to find a way to mimick up the heads and fail cleanly, far enough to let the canResume do its thing
+                    rebuildEmptyGraph();
+                    return;
+                }
             }
+
             headsSerial = null;
             // Same for startNodes:
             startNodes = new Stack<BlockStartNode>();
             for (String id : startNodesSerial) {
-                startNodes.add((BlockStartNode) storage.getNode(id));
+                FlowNode node = storage.getNode(id);
+                if (node != null) {
+                    startNodes.add((BlockStartNode) storage.getNode(id));
+                } else {
+                    // TODO if possible, consider tyring to close out unterminated blocks to keep existing graph history
+                    rebuildEmptyGraph();
+                    return;
+                }
             }
             startNodesSerial = null;
         }
     }
 
+    /** If true, we are allowed to resume the build because resume is enabled AND we shut down cleanly. */
+    public boolean canResume() {
+        if (isResumeBlocked()) {
+            return false;
+        }
+        if (persistedClean != null) {
+            return persistedClean.booleanValue();
+        }
+        FlowDurabilityHint hint = getDurabilityHint();
+        return hint.isPersistWithEveryStep();
+    }
+
     @Override
     public void onLoad(FlowExecutionOwner owner) throws IOException {
         this.owner = owner;
         try {
             initializeStorage();
             try {
                 if (!isComplete()) {
-                    loadProgramAsync(getProgramDataFile());
+                    if (canResume()) {
+                        loadProgramAsync(getProgramDataFile());
+                    } else {
+                        // TODO if possible, consider tyring to close out unterminated blocks to keep existing graph history
+                        // That way we can visualize the graph in some error cases.
+                        LOGGER.log(Level.WARNING, "Pipeline state not properly persisted, cannot resume "+owner.getUrl());
```

Comment:
Not sure we really need a warning in the system log for this, so long as a clear error is printed to the build log. This is a pretty normal condition after all.